bazel/PPROF.md

```diff
@@ -18,80 +16,75 @@ Build the static binary using bazel:
 
     $ bazel build //source/exe:envoy-static
 
-### Running a statically-linked Envoy with `pprof`
+### Collecting the profile
 
-And run the binary with a `HEAPPROFILE` environment variable, like so:
+To collect a heap profile, run a statically-linked Envoy with `pprof`
 
-    $ HEAPPROFILE=/tmp/mybin.hprof bazel-bin/source/exe/envoy-static <args>
+and run the binary with a `CPUPROFILE` or `HEAPPROFILE` environment variable, like so:
 
-`HEAPPROFILE` sets a location for the profiler output. A statically-linked
-binary must be run with this environment variable; a dynamically-linked binary
-will populate the working directory by default. (See *Methodology*.)
+    $ CPUPROFILE=/tmp/mybin.cpuprof bazel-bin/source/exe/envoy-static <args>
+    $ HEAPPROFILE=/tmp/mybin.heapprof bazel-bin/source/exe/envoy-static <args>
 
-## Dynamic Linking
+`CPUPROFILE` or `HEAPPROFILE` sets a location for the profiler output. (See *Methodology*.)
 
-### Adding `tcmalloc_dep` to Envoy
+There are several other environment variables that can be set to tweak the behavior of gperftools. See https://gperftools.github.io/gperftools/ for more details.
 
-A statically-linked Envoy will profile everything. In a dynamically-linked
-Envoy, you must add the HeapProfiler instructions yourself.
-`HeapProfilerStart()` will start recording allocations, `HeapProfilerStop()`
-will stop recording, and `HeapProfilerDump()` will dump an output to the
-specified directory. (See [Gperftools Heap
-Profiler](https://gperftools.github.io/gperftools/heapprofile.html).)
+### Analyzing the profile
 
-To add a `HeapProfiler` breakpoint yourself, add `tcmalloc` as a
-dependency under the `envoy_cc_library` rule:
+[pprof](https://github.com/google/pprof) can be used to symbolize CPU and Heap profiles.  For example:
 
-`source/exe/BUILD`
+    $ pprof -text bazel-bin/source/exe/envoy-static /tmp/mybin.cpuprof
 
-```c++
-    envoy_cc_library(
-       name = "envoy_common_lib",
-+      tcmalloc_dep = 1,
-       deps = [
-       ...
-    )
-```
+## Collecting CPU or HEAP profile for the full execution of a test target
 
-It is then necessary to add `HeapProfilerStart()` and `HeapProfilerDump()`
-breakpoints somewhere in Envoy. One place to start profiling is at the
-instantiation of `MainCommonBase::MainCommonBase`:
+The profiler library is automatically linked into envoy_cc_test targets.
 
-`source/exe/main_common.cc`
+Run a test with CPU profiling enabled, like so:
 
-```c++
-    // includes
-    #include "gperftools/heap-profiler.h"
-    ...
-    MainCommonBase::MainCommonBase(...) : ... {
-+       HeapProfilerStart("main_common_base"); // first line
-        ...
-    }
-```
+    $ bazel test --test_env=CPUPROFILE=/tmp/cpuprof <test target>
```

Comment:
For just CPU profiling, I have had to explicitly wipe out  the HEAP* environment variables to avoid polution: 
(for context, see https://github.com/envoyproxy/nighthawk/commit/dfab764fb678089687d8ddb5419972ea06b14df8)

Also maybe @jmarantz has thoughts on specific build options for CPU profiling here?:

Reply:
I think you have a point about --test_env=HEAPCHECK= in test targets or some other way to disable things like this.  For most test runs we do want to do HEAPCHECK, just not when we're trying to get some performance info.

Reply:
Sure, but for heap profiling I think it's fine to turn this off if it interferes with the profiling.

Reply:
Added a note about that with alternate instructions.  Profiling with heap checking enabled makes some sense in cases where you are trying to understand why a test is slow.